homeassistant/components/xiaomi_miio/number.py

```diff
@@ -0,0 +1,209 @@
+"""Support for Xiaomi Mi Air Purifier and Xiaomi Mi Air Humidifier."""
+from dataclasses import dataclass
+from enum import Enum
+import logging
+
+from homeassistant.components.number import NumberEntity
+from homeassistant.const import CONF_HOST, CONF_TOKEN
+from homeassistant.core import callback
+
+from .const import (
+    CONF_DEVICE,
+    CONF_FLOW_TYPE,
+    CONF_MODEL,
+    DOMAIN,
+    FEATURE_FLAGS_AIRHUMIDIFIER,
+    FEATURE_FLAGS_AIRHUMIDIFIER_CA4,
+    FEATURE_FLAGS_AIRHUMIDIFIER_CA_AND_CB,
+    FEATURE_SET_MOTOR_SPEED,
+    KEY_COORDINATOR,
+    KEY_DEVICE,
+    MODEL_AIRHUMIDIFIER_CA1,
+    MODEL_AIRHUMIDIFIER_CA4,
+    MODEL_AIRHUMIDIFIER_CB1,
+    MODELS_HUMIDIFIER,
+    SERVICE_SET_MOTOR_SPEED,
+)
+from .device import XiaomiCoordinatedMiioEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_MOTOR_SPEED = "motor_speed"
+ATTR_ACTUAL_MOTOR_SPEED = "actual_speed"
+
+SERVICE_TO_METHOD = {
+    SERVICE_SET_MOTOR_SPEED: {
+        "method": "async_set_motor_speed",
+        "property": ATTR_MOTOR_SPEED,
+    },
+}
+
+
+@dataclass
+class NumberType:
+    """Class that holds device specific info for a xiaomi aqara or humidifier number controller types."""
+
+    name: str = None
+    short_name: str = None
+    unit_of_measurement: str = None
+    icon: str = None
+    device_class: str = None
+    min: float = None
+    max: float = None
+    step: float = None
+    service: str = None
+
+
+NUMBER_TYPES = {
+    FEATURE_SET_MOTOR_SPEED: NumberType(
+        name="Motor speed",
+        icon="mdi:fast-forward-outline",
+        short_name=ATTR_MOTOR_SPEED,
+        unit_of_measurement="rpm",
+        min=200,
+        max=2000,
+        step=10,
+        service=SERVICE_SET_MOTOR_SPEED,
+    ),
+}
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Selectors from a config entry."""
+    entities = []
+    if config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
+        host = config_entry.data[CONF_HOST]
+        token = config_entry.data[CONF_TOKEN]
+        model = config_entry.data[CONF_MODEL]
+        device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+        coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+        device_features = 0
+        entity_class = None
+
+        _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
```

Comment:
IMO we should log about device init only once [here](https://github.com/jbouwh/core/blob/add-xiaomi-miio-humidifier-controls/homeassistant/components/xiaomi_miio/__init__.py#L104)

Reply:
Please review PR https://github.com/home-assistant/core/pull/52173. `number.py` is no longer an effected part of the PR.